### ISSUE TITLE
Finding nodes with unique key

### ISSUE THREAD

**Issue by dsyzling** [Alexander-Miller/treemacs#499]
I've built a treemacs tree of nodes from data provided by an lsp server. The lsp server identifies each node uniquely using a nodeUri. Asynchronous messages will be sent to my extension giving me updates to nodes - the mechanism they use to identify a node is the nodeUri. Whilst I've attached the nodeUri as the key-form I can't easily locate the node when the message arrives.

treemacs-find-node takes a path but I won't receive the path from the lsp server. treemacs-find-in-dom does take a key although I think you suggested previously to not use this function as it was internal and could be unreliable/open to change in future. 

The best I could do is to potentially build my own index of nodeUri -> full path expressions by querying the parent path within my query-function and for each child calculating their full path and building the map. I'd also have to carefully track nodes being deleted and moved. 

Has this situation arisen before with other extension use cases or has this feature been requested in the past to find nodes with a unique key?

**Comment by Alexander-Miller** [comment]
Actually I ended up with a node's unique path being constructed from its and its parents' more or less unique keys because I thought it'd make things easier, otherwise users would have to ensure that every single node key is truly unique, and I didn't know how difficult that'd really be. And now you come along and put the whole thing on its head. 

So yeah, it's a new thing and after thinking about it now I see no way to make it work with the current api. I think I'll need to introduce some kind of `:unique-key-form` keyword thal'll directly be used as a path instead of consing it with the parent.

That'll take a day or two, if you need an interim solution I think you can try manually finding a node where the nodeUri is the last element of its path.

**Comment by dsyzling** [comment]
I guess the problem is the documentation for treemacs-render-node says that key-form is a unique key :-) I understand you may have backward compatibility issues if other projects haven't generated a unique value and they could break in subtle and not so subtle ways.

A day or two would be fine for me - but absolutely no rush, you've been far more responsive than I have any right to expect.

In case I need an interim - or do understand the api a little better - what would be the supported ways for exploring the tree root through to leaves?

**Comment by Alexander-Miller** [comment]
Wait, no, this isn't going to work. Treemacs actually needs a fully qualified node path for its navigation. Basically if you want to move to node (a b c) treemacs will look up the full path in the dom. If that doesn't work it'll look for the node (a b) and then find (a b c) with a manual scan among its children, and so on until we start with manually scanning the whole path from the project root. 

So even if you have a properly unique key for every node that still breaks navigation, or, at best, requires a complete top-down scan for the key.

**Comment by Alexander-Miller** [comment]
> In case I need an interim - or do understand the api a little better - what would be the supported ways for exploring the tree root through to leaves?

That depends. If  you can assume that whatever you're looking for will be an expanded directory then it's in the dom. Then you could find the root node in the dom and navigate via its children (the `treemacs-dom-node->` functions should be enough then)

If not then I see little choice than scanning every single line and checking its `path`. That may not be too horrible if your buffer is small enough (it's still horrible though).

I have considered the option of expanding the internal dom to include the leaves, but I'll have to see how it impacts performance, and whether it makes the rest of the code base easier or more difficult to handle.

**Comment by dsyzling** [comment]
What would be the overhead of building an index of unique-keys -> path definitions of nodes that have been loaded in the tree. I don't see a need for finding a node that hasn't been loaded yet - only nodes that the tree is aware of that have been loaded through the query-function etc. If find-node or similar could use a unique key, map to a path, tree navigation would work. The downside would be maintaining such an index when nodes (and their children) are moved/deleted etc.  If an entire subtree were moved potentially this could trigger path changes for all sub nodes in the index (depending on what data structure was used). 

I'm also going to take another look at the Metals nodeUri format, I'm wondering if I can glean the treemacs path from the structure of uri - although there's nothing in the spec to suggest there's any relation between the uri string value and it's position within a hierarchy (i.e. parent URIs).

**Comment by Alexander-Miller** [comment]
Assuming repeated lookups dominate inserts and deletions you could
handle your own position caching like this:

- Define a table mapping nodeUris to nodes (markers).
- Whenever you need to find a node given by its nodeUri you don't go
through treemacs' api, but look at your table first.
- Recompute if the value from the table is either nil, or the cached value
does not match the given nodeUri because something moved too much in the
meantime:
(equal (-last (treemacs-button-get (gethash nodeUri table) :path)) nodeUri)
- Recomputation happens with the dumb top-down lookup: move through
every line and check if the node at point has the right nodeUri in
its path


It's not ideal, but it gets the job done for now.

**Comment by dsyzling** [comment]
Thanks - yes I've prototyped something that builds the cache index during the query-children function by check the current node's path - I can then build the path of the child node being created by adding it's 'key' in my case nodeUri. I need to deal with the situation where the cached value no longer matches. 

One element I'm confused about is if the node's properties change in some way - in one of my cases the label changes. The way I've handled this so far is attaching my original data item to the node in an extended property. I then update the value of the label in this extended data item. My label render form would normally extract this property to display the label when the parent is expanded. 

If I receive an update, I find the nodeUri in my index, retrieve the path and locate it using find-node, swap the label over in my item but now I want to force the label to show its new value. Does the label form only get called during treemacs-render-node and therefore do I have to retrieve the parent node of the one I'm updating and request that this node is updated using treemacs-do-update-node or treemacs-update-node? If I try and update the node itself I don't see the display update, only when I use (treemacs-do-update-node (treemacs-parent-of tree-node) nil) do I see the tree update - I presumed this was because this was forcing the treemacs-render-node call which would then trigger my label-form to be evaluated and the display updated.

I just wanted to make sure I was approaching updating the tree in place correctly and not missing something obvious. Equally I didn't necessarily want to add more unnecessary processing overhead than I needed to.

**Comment by dsyzling** [comment]
The drawback with this approach I'm finding though - is that if you attempt to interact with the tree - for example I might want to expand the node that's being updated because children are being displayed with compilation status - updating the parent causes the tree to refresh the cursor to track to the root node - so I'm constantly moving the cursor to the node to expand and it's being reset back to the root because of the update.

**Comment by Alexander-Miller** [comment]
> Does the label form only get called during treemacs-render-node

I'd have to expand some macros to be sure, but probably yes.

> only when I use (treemacs-do-update-node (treemacs-parent-of tree-node) nil) do I see the tree update

That's the way it's set up to work. What an update-node call does is basically move to the node and press TAB twice (if the node is TABable and in an expanded state). So it's not really so much about updating an actual node, but rather its content.

A function to rename a node is very high up on my agenda, so you can soon stop reloading entire subtrees just for a new label.

> updating the parent causes the tree to refresh the cursor to track to the root node

That shouldn't happen. There's usually two different version of updating functions called, for example, `update-node` and `do-update-node`. The former is just a call to the latter using the `treemacs-save-position` macro (that has some overhead, hence the separation). That macro should be smart enough to either keep point on the same node (same `:path`), even if the update made it move, or, if the node cannot be found anymore, to at least stick with some neighbors or parents.

So first make sure you're calling the right things. If you are then we're dealing with a bug. One that would be very difficult to track down on your end (you would need some deeper understanding of treemacs' navigation capabilities), so the best way to handle this would be some minimal example that I can reproduce the bug with.

**Comment by dsyzling** [comment]
Yes I understand it's difficult, on my side I have asynchronous messages arriving and then I'm sending async responses which makes testing difficult and creating an isolated test case equally hard. The Metals treeview protocol isn't without bugs at the moment and there's inconsistent behaviour under vscode.

I think I've managed to track down the issue - I added my own save-excursion to avoid the focus being set back to the treeview when updates occur when you're editing - I think this is conflicting with the treemacs-update call. So I removed my save-excursion and reverted to using treemacs-update - rather than do-update. I think this may have resolved a couple of issues, but need to do some further more detailed testing.
